### DISCUSSION TITLE
Why are the results from the NPR Specular node and Toon Shading node so different?

### DISCUSSION THREAD

**Discussion by dibli-goost** [bnpr/Malt#400] [Q&A]
Finally getting around to converting my shaders that used the Toon Shading node to the NPR Specular one, and man... I can't for the life of me get a similar result from the two.

Toon Shading (set to specular of course):
![image](https://user-images.githubusercontent.com/31767887/192348490-94371e03-8ce7-4e5c-b54b-187480aa9477.png)

NPR Specular:
![image](https://user-images.githubusercontent.com/31767887/192348585-5de434dd-efea-47a6-b6c9-ba37bb072afe.png)

They both have the same general settings here:
![image](https://user-images.githubusercontent.com/31767887/192348802-696af6d0-7feb-4906-95dd-3182289a5478.png)

I had been using a distance node to try and tease the same output them, but I just can't do it. What is going on here? Are they simply using different specular models? This is the closest I can get to them canceling out:

![image](https://user-images.githubusercontent.com/31767887/192350214-de6ce02d-9c8c-4367-b852-1320a61e506d.png)

I much prefer the look of the older Toon Shading node here, especially when using custom normals.
NPR Specular Left, Toon Shading right:
![image](https://user-images.githubusercontent.com/31767887/192351561-ca436f67-07e8-4648-8905-af78190c11d2.png)

My normal setup completely overrides the normals of the face, axing details like the nose, but for whatever reason the NPR specular node is reintroducing those details. The result IS different from the unedited normals, but idk.. It looks right using the Toon Shading node.

**Comment by pragma37** [comment]
Yes, they are just different shading models.
The old toon shader uses the Phong model and the new one is a tweaked GGX.

I tried to find a model that mapped the specular falloff in a smooth and linear way so it's easier to remap using a color ramp.
Here I managed to get a somewhat close result using a color ramp:
![imagen](https://user-images.githubusercontent.com/36610999/192572426-c85d3e82-16b2-4108-829e-dbd97061a267.png)
*(Color Ramp at the left and old Toon Shader at the right)*

But yeah... it's not perfect.
The ideal solution would be to allow remapping the specular falloff with a vector curve UI, so you can get the exact result you want.
You can already do that with a color ramp, but it's harder to control.

**Reply by pragma37** [reply]
That custom normal issue seems like something we should fix, though.  
If you can send me the head model I can take a look at it.

**Reply by dibli-goost** [reply]
> Yes, they are just different shading models.
The old toon shader uses the Phong model and the new one is a tweaked GGX.

What about an option to swap between models on the new node? 

> That custom normal issue seems like something we should fix, though. If you can send me the head model I can take a look at it.

[Cuistom Vcol Normals issue.zip](https://github.com/bnpr/Malt/files/9659088/Cuistom.Vcol.Normals.issue.zip)

Normals are generated via a geometry nodes setup, and sent to Malt via Vertex Colors. Made a basic set up with Diffuse, Specular and the Toon Specular to inspect them.

(Side note: I have a second material on the head, with a node tree that acts funny with color properties, for some reason it changes them to the standard 3/4 vector inputs rather than the usual color inputs for me. No idea why it started doing that. But yeah it's there 🤷‍♀️)

**Reply by pragma37** [reply]
Seems like the "nose normals" issue is caused by the tangents/bitangents, which are needed for supporting anisotropy.
I need to look into it with more time.

> What about an option to swap between models on the new node?

The point is that it should be possible to replicate other models with the color ramp and eventually with vector curves, so I'm not too sold on the idea.

For now, you can drop this function into a plugin. It's the old Toon shading model with the new node options:

```glsl
#if defined(IS_MESH_SHADER) || defined(IS_SCREEN_SHADER)

#ifdef IS_MESH_SHADER
/*  META
    @meta: label=Color Layer Toon; subcategory=NPR Specular;
    @base_color: default=vec3(0);
    @color: default=vec3(1);
    @size: subtype=Slider; min=-0.0; max=1.0; default=1.0;
    @gradient_size: subtype=Slider; min=0.0; max=1.0; default=0.1;
    @position: subtype=Vector; default=POSITION;
    @normal: subtype=Normal; default=NORMAL;
    @tangent: subtype=Normal; default=radial_tangent(NORMAL, vec3(0,0,1));
    @offset: subtype=Slider; min=-1.0; max=1.0; default=0.0;
    @light_groups: default=MATERIAL_LIGHT_GROUPS;
    @shadows: subtype=ENUM(Inherit from Material,Enable Shadows,Disable Self-Shadows,Disable Shadows); default=0;
*/
#else
/*  META
    @meta: label=Color Layer Toon; subcategory=NPR Specular;
    @base_color: default=vec3(0);
    @color: default=vec3(1);
    @size: subtype=Slider; min=-0.0; max=1.0; default=1.0;
    @gradient_size: subtype=Slider; min=0.0; max=1.0; default=0.1;
    @position: subtype=Vector; default=POSITION;
    @normal: subtype=Normal; default=NORMAL;
    @tangent: subtype=Normal; default=radial_tangent(NORMAL, vec3(0,0,1));
    @offset: subtype=Slider; min=-1.0; max=1.0; default=0.0;
    @light_groups: default=ivec4(1,0,0,0);
    @shadows: subtype=ENUM(Enable Shadows,Disable Self-Shadows,Disable Shadows); default=0;
*/
#endif
vec3 NPR_specular_layer(
    vec3 base_color,
    vec3 color,
    float size,
    float gradient_size,
    float offset,
    bool max_contribution,
    int shadows,
    ivec4 light_groups,
    vec3 position,
    vec3 normal,
    vec3 tangent
)
{
    bool shadow;
    bool self_shadow;
    _shadow_params(shadows, shadow, self_shadow);

    size = saturate(size + offset) * 0.99;
    gradient_size = min(size, gradient_size);
    
    vec3 result = vec3(0,0,0);
    for (int i = 0; i < LIGHTS.lights_count; i++)
    {
        for(int group_index = 0; group_index < 4; group_index++)
        {
            if(LIGHT_GROUP_INDEX(i) != light_groups[group_index])
            {
                continue;
            }
            Light L = LIGHTS.lights[i];
            LitSurface LS = npr_lit_surface(position, normal, ID.x, L, i, shadow, self_shadow);

            if(LS.NoL < 0)
            {
                continue;
            }

            float VoR = dot(LS.V, LS.R);
            float angle = acos(VoR);
            float S = 1.0 - (angle / (PI/2));
            
            //minimal geometric shadowing
            S *= 1.0 - pow(1.0 - LS.NoL, 5);

            vec3 specular = _map_gradient(size, gradient_size, vec3(S)) * LS.light_color * LS.shadow_multiply;
            
            if(max_contribution)
            {
                result = max(result, specular);
            }
            else
            {
                result += specular;
            }
        }
    }
    
    return mix(base_color, color, result);
}

#endif
```

I looked into the broken node tree.  
It fails because there's a property override for a property that doesn't exist.  

I have no idea how that could have happened.  

It would be easy to "patch", but I fear it may hide the root issue.  
If you find a way to reproduce the issue I can take a look into it.
Creating a new tree and copy-pasting the nodes should do the trick for now.

**Reply by dibli-goost** [reply]
> Seems like the "nose normals" issue is caused by the tangents/bitangents, which are needed for supporting anisotropy. I need to look into it with more time.

Hmm, this setup is based off of the one Ascalon tutorialized in his recent tutorial series. It might not be an issue with Malt itself, but rather that I need a generated Custom Tangent as well in order for the node to function properly. I've asked them about it on Discord in the meantime.
 
> For now, you can drop this function into a plugin. It's the old Toon shading model with the new node options:

Aight, I got it working. I'll use this in the mean time.

> If you find a way to reproduce the issue I can take a look into it. Creating a new tree and copy-pasting the nodes should do the trick for now.

I'll keep an eye out for if it happens again. 👍

**Reply by aVersionOfReality** [reply]
Any situation where you need Tangents that are Orthogonal to the Normal will get messed up by Custom Normals of any sort. I've run into it with Normal Mapping mostly. Is the Specular doing something with the Tangent even when the Anisotropy is disabled? (Is it disabled?)

Currently, the only option I'm aware of to get Custom UV Tangents to go with Custom Normals is to make both in GeoNodes (which requires a custom UV tangent group that only I have, lol. We really need a proper UV tangent node in GN.) You cannot make a Tangent from a Normal in the shader as far as I'm aware. This is because UV Tangent calculation is mesh level and is based on edge length ratios between UV space and World space (if I've understood it correctly). And for that you need a For loop. (Which Geonodes doesn't have, but you can fake it with Accumulate node.)

**Reply by pragma37** [reply]
> Is the Specular doing something with the Tangent even when the Anisotropy is disabled? (Is it disabled?)

Ideally, I should find a specular model where the tangents/bitangents are ignored when there's no anisotropy, but the current anisotropic GGX model I use seems to differ from the regular GGX model even when the anisotropy is 0.5.

> You cannot make a Tangent from a Normal in the shader as far as I'm aware.

It's possible to compute an approximation (it works fine for dense meshes).
Malt does that by default: https://github.com/bnpr/Malt/blob/Development/Malt/Shaders/Common/Normal.glsl#L54

But it's computed based on the position derivatives, so it doesn't really solve the issue.
